sycl/doc/design/ESIMDStatelesAccessors.md

```diff
@@ -0,0 +1,165 @@
+
+## Accessor and USM pointer kernel argument passing details
+ESIMD compiler when compiling a kernel records information about each memory
+argument and stores it together with the kernel's SPIRV. Basically, for each
+kernel argument, there is information whether it is a memory argument, and, if
+yes, whether it is surface-based or pointer-based.
+
+When JITting the kernel, the scalar GPU compiler back-end can convert memory
+arguments and memory accesses between the two modes depending on optimization
+or other settings, and record final type of memory argument with the generated
+kernel executable. GPU runtime uses that information to wrap/not wrap incoming
+memory pointer with a surface before passing it onto the harware into the
+actual kernel argument.
+
+The vector back-end can't do this in many cases, as memory accesses in SPIRV
+are represented by hardware-specific intrinsics rather then standard generic
+memory access SPIRV instructions. This design basically enables the vector BE
+to redirect code generation for stateful memory access APIs to stateless or
+stateful intrinsics, and also generate correct annotations. Since it uses the
+same runtime, which relies on parameter annotation when making the wrap/no-wrap
+decision, the runtime part does not need much changes.
+
+## Problem definition
+
+Currently, ESIMD compiler always maps buffer/accessor-based memory accesses to
+stateful accesses, thus imposing the 4Gb datum size limitation on user programs
+with accessors.
+
+## Proposed solution
+
+### Short/mid-term
+
+#### API header changes
+The general idea is to introduce C++ preprocessor macro
+`ESIMD_FORCE_STATELESS_MEM_ACCESS` which will control code generation for the
+stateful memory access APIs - such as:
+
+```cpp
+template <typename Tx, int N, typename AccessorTy,
+          typename Flags = vector_aligned_tag,
+          typename = std::enable_if_t<is_simd_flag_type_v<Flags>>,
+          class T = detail::__raw_t<Tx>>
+__ESIMD_API simd<Tx, N> block_load(AccessorTy acc, uint32_t offset,
+                                   Flags = {});
+
+template <typename T, int N, typename AccessorTy>
+__ESIMD_API std::enable_if_t<(sizeof(T) <= 4) &&
+                                 (N == 1 || N == 8 || N == 16 || N == 32) &&
+                                 !std::is_pointer<AccessorTy>::value,
+                             simd<T, N>>
+gather(AccessorTy acc, simd<uint32_t, N> offsets, uint32_t glob_offset = 0,
+       simd_mask<N> mask = 1);
+```
+
+Implementation of the APIs would follow this pattern:
+
+```cpp
+T stateful_memory_api(accessor acc, uint32 offset, args...) {
+#ifdef ESIMD_FORCE_STATELESS_MEM_ACCESS
+  accessor_elelemt_type *ptr = acc.get_pointer() + offset;
+  return stateless_memory_api(ptr, args...);
+#else
+  <original implementation>
+#endif
+}
+```
+
+The new macro is supposed to be set by users directly or via some logic based on
+other macros set by users.
```

Comment:
@kbobrovs I'm assuming that `ESIMD_FORCE_STATELESS_MEM_ACCESS` must be defined in order to use `accessor::get_pointer()` in ESIMD kernels, right?

Reply:
@petercad - yes. `-fsycl-esimd-force-stateless-mem-access` should ensure this automatically